feat: add /cid/:cid to linkdex CARs by cid

olizilla opened the pull request:
Try and find a set of CARs that contain a complete DAG for a given CID.

For dotStorage CARs are grouped by userid with keys like `raw/<cid>/<userid>/<hash>.car`.

We find all the CARs for a given cid by prefix, group them by userid, and then sort by count of CARs, under the assumptions that the upload with the most CARs is likely to contain the full DAG.

This is needed for our cron pin status job in dotStorage so we can land https://github.com/web3-storage/web3.storage/pull/1785

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

olizilla committed:
feat: add /cid/:cid to linkdex CARs by cid

Try and find a set of CARs that contain a complete DAG for a given CID.

For dotStorage CARs are grouped by userid with keys like `raw/<cid>/<userid>/<hash>.car`.

We find all the CARs for a given cid by prefix, group them by userid, and then sort by count of CARs, under the assumptions that the upload with the most CARs is likely to contain the full DAG.

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

olizilla committed:
fix: 404 for CID we don't have

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

olizilla commented:
This in hold as we don't currently need it. We have the `key` in the dotstroage worker, and after the last s3 upload of a set of partials, the final call as access to all the partial cars. We don't need a cron job to find it later, where we might only have the CID.

This may change if we move to a different s3 key structure.

alanshaw commented on services/functions/linkdex-cid.ts:
```diff
@@ -0,0 +1,71 @@
+  if (!CID.parse(cid)) {
+    return response({ error: { details: 'cid is invalid' }}, { statusCode: 400 })
+  }
+  const prefix = `raw/${cid}/`
```
Not blocking, but we should consider CARs in `complete/${cid}`. For examples, CARs from [backup](https://github.com/web3-storage/backup/blob/main/index.js#L133) are written there.

alanshaw commented on services/functions/linkdex-cid.ts:
```diff
@@ -0,0 +1,71 @@
+    await pMap(carKeys, indexer, { concurrency: 10 })
+    if (index.getDagStructureLabel() === 'Complete') {
+      // return as soon as we find a set of CARs that has a complete DAG
+      return response({ structure: 'Complete' })
```
Shouldn't we return the full report here?

alanshaw commented:
https://github.com/web3-storage/linkdex-api/pull/11 addresses the two review comments so provided we can merge it here we should be good to go.

Commit:
feat: report for CARs in the complete directory (#11)

* Introduces a reporter for CARs in the `complete` directory
* Augments the responses to include the full report (instead of just
structure)
* Adds bucket name to CAR key (for CID URL) so you know which bucket the
CARs are in (not wedded to this but I thought it might be useful if we
wanted to use linkdex for the new w3up bucket also)

Can be made a little nicer if we merge+release
https://github.com/web3-storage/linkdex/pull/13

Commit:
Merge branch 'main' into linkdex-by-cid

Commit:
Merge branch 'linkdex-by-cid' of github.com:web3-storage/linkdex-api into linkdex-by-cid

Commit:
fix: function definitions
